### ISSUE TITLE
`BuiltList<SomeType> get someProp;` can't handle `null` values with JSON deserialization

### ISSUE THREAD

**Issue by zoechi** [google/built_value.dart#149]
```dart
abstract class ProfileGeneralDetails extends BaseModel
    implements Built<ProfileGeneralDetails, ProfileGeneralDetailsBuilder> {
  ...
  @nullable
  BuiltList<LocalisedString> get characteristics;
}
```
```dart
abstract class Profile extends BaseModel
    implements Built<Profile, ProfileBuilder> {

  @nullable
  ProfileGeneralDetails get general;
  ...
}
```
```dart
    final profileData = JSON.decode("""
{
"id": 1915021,
"general": {
    "characteristics": null
},
```

causes

```
Unhandled exception:
NoSuchMethodError: The method 'map' was called on null.
Receiver: null
Tried calling: map(Closure: (dynamic) => dynamic)
#0      Object._noSuchMethod (dart:core-patch/object_patch.dart:44)
#1      Object.noSuchMethod (dart:core-patch/object_patch.dart:47)
#2      BuiltListSerializer.deserialize (package:built_value/src/built_list_serializer.dart:44:30)
#3      BuiltJsonSerializers._deserialize (package:built_value/src/built_json_serializers.dart:139:27)
#4      BuiltJsonSerializers.deserialize (package:built_value/src/built_json_serializers.dart:103:18)
#5      _$ProfileGeneralDetailsSerializer.deserialize (package:myproj/src/models/profile_general_details.g.dart:121:54)
#6      BuiltJsonSerializers._deserialize (package:built_value/src/built_json_serializers.dart:139:27)
#7      BuiltJsonSerializers.deserialize (package:built_value/src/built_json_serializers.dart:103:18)
#8      _$ProfileSerializer.deserialize (package:myproj/src/models/profile.g.dart:154:46)
#9      BuiltJsonSerializers._deserialize (package:built_value/src/built_json_serializers.dart:139:27)
#10     BuiltJsonSerializers.deserialize (package:built_value/src/built_json_serializers.dart:103:18)
#11     Profile.Profile.fromJsonMap (package:myproj/src/models/profile.dart:31:24)
#12     main.<anonymous closure> (file:///home/zoechi/dart/customers/mycust/myproj.dart/bin/models_test.dart:246:25)
#13     Declarer.group.<anonymous closure> (package:test/src/backend/declarer.dart:144:37)
#14     _rootRun (dart:async/zone.dart:1150)
#15     _CustomZone.run (dart:async/zone.dart:1026)
#16     runZoned (dart:async/zone.dart:1503)
#17     Declarer.declare (package:test/src/backend/declarer.dart:96:22)
#18     Declarer.group (package:test/src/backend/declarer.dart:141:14)
#19     group (package:test/test.dart:200:13)
#20     main (file:///home/myuser/dart/customers/mycust/myproj.dart/bin/models_test.dart:12:3)
<asynchronous suspension>
#21     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:261)
#22     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:148)
```

**Comment by zoechi** [comment]
Similar with a `double` property with value `null`
```
  NoSuchMethodError: The method 'toDouble' was called on null.
  Receiver: null
  Tried calling: toDouble()
  dart:core                                                    Object.noSuchMethod
  package:built_value/src/num_serializer.dart 40:34            NumSerializer.deserialize
  package:built_value/src/built_json_serializers.dart 142:27   BuiltJsonSerializers._deserialize
  package:built_value/src/built_json_serializers.dart 103:18   BuiltJsonSerializers.deserialize
  package:myproj/src/models/profile_summary.g.dart 210:41  _$ProfileSummarySerializer.deserialize
  package:built_value/src/built_json_serializers.dart 139:27   BuiltJsonSerializers._deserialize
  package:built_value/src/built_json_serializers.dart 103:18   BuiltJsonSerializers.deserialize
  package:built_value/src/built_json_serializers.dart 33:12    BuiltJsonSerializers.deserializeWith
  package:myproj/src/models/profile_summary.dart 27:24     ProfileSummary.ProfileSummary.fromJsonMap
  package:myproj/src/models.dart 161:20                    ModelFactory.fromJsonMap
  package:myproj/src/app.dart 102:30                       MyApp._processQueryResponse.<fn>.<fn>
  dart:core                                                    List.forEach
  package:myproj/src/app.dart 100:24                       MyApp._processQueryResponse.<fn>
  dart:collection                                              _HashVMBase&MapMixin&&_LinkedHashMapMixin.forEach
  package:myproj/src/app.dart 70:12                        MyApp._processQueryResponse
  package:myproj/src/app.dart 47:12                        MyApp.query
  test/api_test.dart 65:32                                     main.<fn>.<fn>
  ===== asynchronous gap ===========================
  dart:async                                                   _Completer.completeError
  package:myproj/src/app.dart 52:3                         MyApp.query
  ===== asynchronous gap ===========================
  dart:async                                                   _asyncThenWrapperHelper
  package:myproj/src/app.dart 40:61                        MyApp.query
  test/api_test.dart 65:32                                     main.<fn>.<fn>
  ===== asynchronous gap ===========================
  dart:async                                                   Future.Future.microtask
  test/api_test.dart 64:56                                     main.<fn>.<fn>
```

for a property

```dart
  @nullable
  num get distance;
```
and value `distance: null`

**Comment by davidmorgan** [comment]
So there are two problems:

1. The deserialization code does not allow nulls, as you have noticed
2. The serialization code does not emit nulls; if a value is null, it just drops the key instead, i.e. it writes nothing

1 is easy to solve, we just ignore nulls. 2 is harder. So my question is: will it be enough to just ignore nulls, or does the serialization also need to output nulls for interop to work?

Thanks!

**Comment by zoechi** [comment]
I don't know about 2, therefore I'd say no, until the requirement actually comes up.

**Comment by zoechi** [comment]
I did some investigation. 
It doesn't seem to be possible to fix that in the `StandardJsonPlugin` alone.

A few changes got me unblocked (with only few tests in application code using `null`)
No idea if this are good places for the changes ...

![image](https://cloud.githubusercontent.com/assets/405837/23988636/58d539da-0a2f-11e7-987a-70e2c8d19b22.png)

![image](https://cloud.githubusercontent.com/assets/405837/23988663/7900a744-0a2f-11e7-95e3-3a7c322b7fa4.png)

![image](https://cloud.githubusercontent.com/assets/405837/23988696/a2ede774-0a2f-11e7-9a79-4269ddb067e2.png)

**Comment by zoechi** [comment]
Did you have a look? 
This is currently a real blocker. Besides this issue, it looks like built_value is a great fit for an Angular2+Flutter project. Switching to default serialization eventually looks like an interesting option for a follow-up project when the server can be made to support it.
Anything I can do to help? I don't understand the code well enough yet, to know what a proper approach would be to fix this.

**Comment by dave26199** [comment]
Sorry, didn't get a chance to take a look -- will try for this week.

My current though is, in StandardJsonPlugin, to look for any null key/value pairs and remove them. Then hopefully you won't need any changes elsewhere. i.e. whenever the plugin encounters a Map, go through all key/values and check them.

If that works then you should be able to add it as your own plugin until it's supported in StandardJsonPlugin.

**Comment by zoechi** [comment]
I see. Thanks a lot for the hint, I'll try that.

**Comment by davidmorgan** [comment]
Released 0.5.7 for this. Please let me know if it fixes the issue for you!

**Comment by zoechi** [comment]
Works fine so far. I'll do more tests soon. Thanks a lot for your super-quick updates :+1:
